### PR TITLE
Adapted policy to the claims format of the IdentityHubCredentialsVerifier.

### PR DESCRIPTION
**Pull request by ouphi**
## What this PR changes/adds

Adapted policy to the claims format returned by the IdentityHubCredentialsVerifier.

## Why it does that

MVD will be using the IdentityHubCredentialsVerifier. Adapted the policy to the claim format that the IdentityHubCredentialsVerifier returns.

## Further notes

Changed the MockCredentialVerifier to return the same format as the IdentityHubCredentialsVerifier.

## Linked Issue(s)

https://github.com/agera-edc/IdentityHub/issues/32

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly?

### CODE REVIEW

File: extensions/mock-credentials-verifier/src/main/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifier.java [hunk 1]
```diff
@@ -66,9 +73,18 @@ public Result<Map<String, Object>> verifyCredentials(String hubBaseUrl, PublicKe
...
                     ));
             monitor.debug("Completing (mock) credential verification. Claims: " + claims);
-            return Result.success(claims);
+            return Result.success(toMappedVerifiableCredentials(claims));
         } catch (MalformedURLException e) {
             throw new EdcException(e);
         }
     }
+
+    private Map<String, Object> toMappedVerifiableCredentials(Map<String, Object> regionClaims) {
```
Comment (ouphi) [inline extensions/mock-credentials-verifier/src/main/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifier.java:93]: This method is duplicated in the tests, but this class MockCredentialsVerifier will be deleted in the next PR so no action needed.

File: extensions/mock-credentials-verifier/build.gradle.kts [hunk 2]
```diff
@@ -14,5 +14,6 @@ dependencies {
     testImplementation("org.assertj:assertj-core:${assertj}")
     testImplementation("org.junit.jupiter:junit-jupiter-api:${jupiterVersion}")
     testImplementation("org.mockito:mockito-core:${mockitoVersion}")
+    implementation("${edcGroup}.identityhub:identity-hub-spi:${edcVersion}")
```
Comment (marcgs) [inline extensions/mock-credentials-verifier/build.gradle.kts]: Variables should be renamed from `edcGroup` and `edcVersion` to `identityHubGroup` and `identityHubVersion`. Specially the versions will deviate at some point.

File: extensions/mock-credentials-verifier/src/test/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifierTest.java [hunk 3]
```diff
@@ -27,20 +27,27 @@
...
     MockCredentialsVerifier verifier = new MockCredentialsVerifier(new ConsoleMonitor());
     PublicKeyWrapper wrapper = mock(PublicKeyWrapper.class);
 
     @Test
     void verifyCredentials() {
         Result<Map<String, Object>> actual = verifier.verifyCredentials("http://dummy.site/foo?region=us&tier=GOLD", wrapper);
         assertThat(actual.succeeded()).isTrue();
-        assertThat(actual.getContent())
-                .isEqualTo(Map.of("region", "us", "tier", "GOLD"));
+        assertThat(extractClaims(actual.getContent())).isEqualTo(Map.of("region", "us", "tier", "GOLD"));
```
Comment (marcgs) [inline extensions/mock-credentials-verifier/src/test/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifierTest.java]: You can use a single assertj assertion here:
```
assertThat(actual.getContent())
                .extracting(c -> c.values().stream().findFirst().get())
                .extracting(VERIFIABLE_CREDENTIAL_KEY)
                .extracting(CREDENTIAL_SUBJECT_KEY)
                .isEqualTo(Map.of("region", "us", "tier", "GOLD"));
```

File: extensions/policies/build.gradle.kts [hunk 4]
```diff
@@ -1,11 +1,33 @@
...
+val jupiterVersion: String by project
+val assertj: String by project
 
 dependencies {
     api("${edcGroup}:ids-spi:${edcVersion}")
     api("${edcGroup}:contract-spi:${edcVersion}")
-}
\ No newline at end of file
+    api("${edcGroup}:core-base:${edcVersion}")
+    implementation("${edcGroup}.identityhub:identity-hub-spi:${edcVersion}")
```
Comment (marcgs) [inline extensions/policies/build.gradle.kts]: Use separate `identityHubVersion` and `identityHubGroup` here

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import org.eclipse.dataspaceconnector.spi.policy.AtomicConstraintFunction;
+import org.eclipse.dataspaceconnector.spi.policy.PolicyContext;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class RegionConstraintFunction implements AtomicConstraintFunction<Permission> {
+    private static final String VERIFIABLE_CREDENTIAL_KEY = "vc";
```
Comment (marcgs) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: This constant is defined centrally at `VerifiableCredentialsJwtService` in the Identity Hub SPI. I would try to use that value instead.

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        var regions = getRegions(context.getParticipantAgent().getClaims());
+        switch (operator) {
+            case EQ:
+                return regions.contains(rightValue);
+            default:
+                return false;
+        }
+    }
+
+    List<String> getRegions(Map<String, Object> claims) {
```
Comment (marcgs) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: private?

File: extensions/mock-credentials-verifier/src/main/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifierExtension.java [hunk 7]
```diff
@@ -15,14 +15,10 @@
...
 import org.eclipse.dataspaceconnector.iam.did.spi.credentials.CredentialsVerifier;
-import org.eclipse.dataspaceconnector.spi.EdcException;
 import org.eclipse.dataspaceconnector.spi.system.Provides;
 import org.eclipse.dataspaceconnector.spi.system.ServiceExtension;
 import org.eclipse.dataspaceconnector.spi.system.ServiceExtensionContext;
 
-import static java.lang.String.format;
-import static org.eclipse.dataspaceconnector.iam.did.spi.document.DidConstants.DID_URL_SETTING;
-
 /**
```
Comment (paullatzelsperger) [inline extensions/mock-credentials-verifier/src/main/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifierExtension.java:24]: use `@Provider` method for the `CredentialsVerifier`

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        } catch (Exception e) {
+            monitor.warning("Error getting verifiable credentials", e);
+            return Optional.empty();
+        }
+    }
+
+    private Optional<String> getRegion(VerifiableCredential verifiableCredential) {
+        try {
+            var region = verifiableCredential.getCredentialSubject().get(REGION_KEY);
+            return region == null ? Optional.empty() : Optional.of((String) region);
```
Comment (marcgs) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: This can be simplified with: `return Optional.ofNullable((String) region);`

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.policy.AtomicConstraintFunction;
+import org.eclipse.dataspaceconnector.spi.policy.PolicyContext;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class RegionConstraintFunction implements AtomicConstraintFunction<Permission> {
```
Comment (paullatzelsperger) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java:34]: should probably be a `AtomicConstraintFunction<Duty>` if we want to enforce a geographic region

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    public RegionConstraintFunction(ObjectMapper objectMapper, Monitor monitor) {
+        this.objectMapper = objectMapper;
+        this.monitor = monitor;
+    }
+
+    @Override
+    public boolean evaluate(Operator operator, Object rightValue, Permission rule, PolicyContext context) {
+        var regions = getRegions(context.getParticipantAgent().getClaims());
+        switch (operator) {
+            case EQ:
```
Comment (paullatzelsperger) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java:48]: why no other operators?

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                return regions.contains(rightValue);
+            default:
+                return false;
+        }
+    }
+
+    List<String> getRegions(Map<String, Object> claims) {
+        return claims.values()
+                .stream().flatMap(x -> getVerifiableCredential(x).stream())
```
Comment (marcgs) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: nit:
```
claims.values()
                .stream()
                .map(o -> (Map<String, Object>)o)
                .flatMap(x -> getVerifiableCredential(x).stream())
                .flatMap(vc -> getRegion(vc).stream())
                .collect(Collectors.toList());
```

this allows then to have `getVerifiableCredential(Map<String, Object> vcObject)` in the method below

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/SeedPoliciesExtension.java [hunk 12]
```diff
@@ -58,7 +66,7 @@ public void initialize(ServiceExtensionContext context) {
         ruleBindingRegistry.bind("USE", ContractDefinitionService.NEGOTIATION_SCOPE);
         ruleBindingRegistry.bind(ABS_SPATIAL_POSITION, ContractDefinitionService.NEGOTIATION_SCOPE);
 
-        policyEngine.registerFunction(ALL_SCOPES, Permission.class, ABS_SPATIAL_POSITION, new AbsSpatialPositionConstraintFunction());
+        policyEngine.registerFunction(ALL_SCOPES, Permission.class, ABS_SPATIAL_POSITION, new RegionConstraintFunction(typeManager.getMapper(), monitor));
```
Comment (paullatzelsperger) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/SeedPoliciesExtension.java:69]: should probably be a `Duty`.

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java (hunk 11 above)
Comment (ouphi) [reply extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: I prefer to keep the casting in a method to catch the exception at the right moment of the flow.

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java (hunk 9 above)
Comment (ouphi) [reply extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java:34]: Are you sure? the [doc](https://www.w3.org/TR/odrl-model/#permission) says "A Permission allows an action, with all refinements satisfied, to be exercised on an Asset if all constraints are satisfied and if all duties are fulfilled."

It corresponds to our use case, here we want to give the permission to access an asset if the condition is fulfilled.

File: extensions/mock-credentials-verifier/src/test/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifierTest.java [hunk 13]
```diff
@@ -43,4 +48,10 @@ void verifyCredentials_failsOnMalformedUrl() {
         assertThatThrownBy(() -> verifier.verifyCredentials("malformed_url", wrapper))
                 .isInstanceOf(EdcException.class);
     }
+
+    private Map<String, Object> extractClaims(Map<String, Object> verifiableCredentials) {
```
Comment (marcgs) [inline extensions/mock-credentials-verifier/src/test/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifierTest.java]: This method is unused

**Review by marcgs — APPROVED** [review]

File: extensions/mock-credentials-verifier/build.gradle.kts [hunk 14]
```diff
@@ -7,12 +7,15 @@ val edcGroup: String by project
 val jupiterVersion: String by project
 val mockitoVersion: String by project
 val assertj: String by project
+val identityHubGroup: String by project
+val identityHubVersion: String by project
 
 dependencies {
     implementation("${edcGroup}:identity-did-spi:${edcVersion}")
```
Comment (marcgs) [inline extensions/mock-credentials-verifier/build.gradle.kts:14]: Could you change this one to `indentityHubVersion` too?

Comment (ouphi) [reply extensions/mock-credentials-verifier/build.gradle.kts:14]: identity-did-spi comes from EDC

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+import org.eclipse.dataspaceconnector.spi.policy.PolicyContext;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.eclipse.dataspaceconnector.identityhub.credentials.VerifiableCredentialsJwtService.VERIFIABLE_CREDENTIALS_KEY;
+
+public class RegionConstraintFunction implements AtomicConstraintFunction<Permission> {
```
Comment (algattik) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java:34]: nit: using git move rather than delete and add the ConstraintFunction would have made changes clearer (and reduced the amount of comments)

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    private List<String> getRegions(Map<String, Object> claims) {
+        return claims.values()
+                .stream().flatMap(o -> getVerifiableCredential(o).stream())
+                .flatMap(vc -> getRegion(vc).stream()).collect(Collectors.toList());
+    }
+
+    private Optional<VerifiableCredential> getVerifiableCredential(Object object) {
```
Comment (algattik) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: based on previous feedback it seems the use of Optional is not allowed

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+        try {
+            var vcObject = (Map<String, Object>) object;
+            var verifiableCredentialMap = vcObject.get(VERIFIABLE_CREDENTIALS_KEY);
+            return Optional.of(objectMapper.convertValue(verifiableCredentialMap, VerifiableCredential.class));
+        } catch (Exception e) {
```
Comment (algattik) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: why such catch-all blocks? This is typically an anti pattern

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java (hunk 16 above)
Comment (ouphi) [reply extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: Do you mean not allowed in EDC? I saw it used in a few places, and here it allows to use flatMap instead of filter(result -> result.successful()).map(result -> result.getContent())

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java (hunk 17 above)
Comment (ouphi) [reply extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: in case of exception during the cast, I can make it clearer with ClassCastException

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java (hunk 16 above)
Comment (algattik) [reply extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: Yeah I of course also prefer it. @paullatzelsperger ?

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Override
+    public boolean evaluate(Operator operator, Object rightValue, Permission rule, PolicyContext context) {
+        var regions = getRegions(context.getParticipantAgent().getClaims());
+        switch (operator) {
+            case EQ:
+                return regions.contains(rightValue);
+            case NEQ:
+                return !regions.contains(rightValue);
+            default:
```
Comment (paullatzelsperger) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java:54]: the `IN` operator would be a quick win here as well

File: extensions/mock-credentials-verifier/src/main/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifier.java [hunk 19]
```diff
@@ -24,37 +24,55 @@
...
      */
     @Override
-    public Result<Map<String, Object>> verifyCredentials(String hubBaseUrl, PublicKeyWrapper othersPublicKey) {
-        monitor.debug("Starting (mock) credential verification against hub URL " + hubBaseUrl);
+    public Result<Map<String, Object>> getVerifiedCredentials(DidDocument didDocument) {
+        monitor.debug("Starting (mock) credential verification from DID document " + didDocument.getId());
+
+        var hubBaseUrlResult = getIdentityHubBaseUrl(didDocument);
+        if (hubBaseUrlResult.failed()) {
+            monitor.debug("Failed to get Hub URL from document");
```
Comment (paullatzelsperger) [inline extensions/mock-credentials-verifier/src/main/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifier.java]: just use `hubBaseUrlResult.getFailureMessage()`

**Review by paullatzelsperger — APPROVED** [review]

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java (hunk 16 above)
Comment (paullatzelsperger) [reply extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: > Yeah I of course also prefer it. @paullatzelsperger ?

`Optional` is ok in fluent blocks, please avoid it as class members, return types or method parameters. Here, returning `null` would be an option, and use `ofNullable(...)` in the caller.

File: .github/actions/gradle-setup/action.yml [hunk 20]
```diff
@@ -9,7 +9,7 @@ runs:
       with:
         repository: eclipse-dataspaceconnector/DataSpaceConnector
         path: DataSpaceConnector
-        ref: 3ff940b720f44826df28e893fb31344eb6faacef
+        ref: v0.0.1-milestone-5
```
Comment (algattik) [inline .github/actions/gradle-setup/action.yml:12]: this PR should block until https://github.com/eclipse-dataspaceconnector/MinimumViableDataspace/pull/34 is merged, or there will be other issues

**Review by algattik — APPROVED** [review]

File: extensions/mock-credentials-verifier/src/main/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifier.java [hunk 21]
```diff
@@ -66,9 +84,29 @@ public Result<Map<String, Object>> verifyCredentials(String hubBaseUrl, PublicKe
...
         }
     }
+
+    private Map<String, Object> toMappedVerifiableCredentials(Map<String, Object> regionClaims) {
+        var vcId = UUID.randomUUID().toString();
+        return Map.of(vcId,
+                Map.of(VERIFIABLE_CREDENTIALS_KEY, Map.of(CREDENTIAL_SUBJECT_KEY, regionClaims,
+                                VERIFIABLE_CREDENTIAL_ID_KEY, vcId),
+                        // issuer will be ignored when applying policies for now.
+                        ISSUER_KEY, String.join("did:web:", UUID.randomUUID().toString())));
```
Comment (cpeeyush) [inline extensions/mock-credentials-verifier/src/main/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifier.java]: nit: just for readability. Up to you to decide. Thanks
```suggestion
        return Map.of(vcId,
                Map.of(VERIFIABLE_CREDENTIALS_KEY,
                        Map.of(CREDENTIAL_SUBJECT_KEY, regionClaims,
                                VERIFIABLE_CREDENTIAL_ID_KEY, vcId
                        ),
                        // issuer will be ignored when applying policies for now.
                        ISSUER_KEY, String.join("did:web:", UUID.randomUUID().toString())
                )
        );
```

File: extensions/mock-credentials-verifier/src/test/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifierTest.java [hunk 22]
```diff
@@ -14,33 +14,47 @@
...
+                .extracting(c -> c.values().stream().findFirst().get())
+                .extracting(VERIFIABLE_CREDENTIALS_KEY)
+                .extracting(CREDENTIAL_SUBJECT_KEY)
                 .isEqualTo(Map.of("region", "us", "tier", "GOLD"));
     }
 
     @Test
     void verifyCredentials_failsOnMalformedUrl() {
-        assertThatThrownBy(() -> verifier.verifyCredentials("malformed_url", wrapper))
+        assertThatThrownBy(() -> getVerifiedCredentials("malformed_url"))
```
Comment (cpeeyush) [inline extensions/mock-credentials-verifier/src/test/java/org/eclipse/dataspaceconnector/mvd/MockCredentialsVerifierTest.java:49]: nit: one more test can be added where DID document is empty. For example, it doesn't have identityHubUrl. That will cause a Result.failure response.

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+            monitor.warning("Error getting verifiable credentials", e);
+            return Optional.empty();
+        }
+    }
+
+    private Optional<String> getRegion(VerifiableCredential verifiableCredential) {
+        try {
+            var region = verifiableCredential.getCredentialSubject().get(REGION_KEY);
+            return Optional.ofNullable((String) region);
+        } catch (Exception e) {
```
Comment (cpeeyush) [inline extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: nit: is there a particular reason why are catching all of the exception here and then returning empty optional, instead not letting exception be thrown or bubble up?

same as in above method `getVerifiableCredential` because this method may also throw IllegalArgumentException and in case of other exceptions current implementation will give the impression that `evaluate` failed with false response but it is actually something else.

Comment (ouphi) [reply extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: I changed the exception to ClassCastException to make it better.
It is in case a region can't be casted to string. To avoid breaking the entire flow checking the policy.

**Review by cpeeyush — APPROVED** [review]
Approved with suggestions.

File: extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java (hunk 23 above)
Comment (cpeeyush) [reply extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java]: or maybe I was thinking too much your trick with optional may also work if you catch all exception with logging error about it.

And just a java doc on the method that we continue to process to avoid any Verifiable credential which do not have region as per the correct format.
